src/main/java/com/google/api/generator/engine/ast/TypeNode.java

```diff
@@ -138,6 +158,24 @@ public int hashCode() {
     return hash;
   }
 
+  private boolean strictEquals(TypeNode other) {
+    return typeKind().equals(other.typeKind())
+        && (isArray() == other.isArray())
+        && Objects.equals(reference(), other.reference());
+  }
+
+  private boolean isBoxedTypeEquals(TypeNode other) {
+    // If both types are primitive/reference type, return false.
+    // Array of boxed/primitive type is not considered equal.
+    if (isPrimitiveType() == other.isPrimitiveType() || isArray() || other.isArray()) {
```

Comment:
We probably don't need the `isPrimitiveType` check, since the map equality check will take care of that.

Reply:
Removed the duplicate check and the comments above. thanks for catching this.